### PR TITLE
bpo-31861: Provide aiter and anext builtins

### PR DESCRIPTION
**Pull request by jab**
This is the C implementation for [bpo-31861](https://bugs.python.org/issue31861) requested as an alternative to the Python implementation I provided in #8895.

For a more direct translation of this into Python (in case it makes reviewing easier), see https://github.com/jab/cpython/commit/ce35092a898be5da4fd6ff2d1c92b4e020ccf934.

Patch by @justin39, @lordmauve, and me.

https://bugs.python.org/issue31861

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@justin39

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Objects/iterobject.c
Comment (eric-wieser) [inline Objects/iterobject.c]: I think you can just pass `value = NULL` here and it will handle building the object if and when needed? Or is there something special about AsyncStopIteration?

Comment (justin39) [reply Objects/iterobject.c]: Hm, would it be easier to just use `PyErr_SetNone` then? Makes sense to me though - I think originally we were setting a value attribute here but got rid of it and never changed how we were creating the exception.

Comment (eric-wieser) [reply Objects/iterobject.c]: Yes, I just couldn't remember the name of `SetNone`

Comment (jab) [reply Objects/iterobject.c]: Thanks! Done in the latest revision.

File: Doc/library/functions.rst
Comment (jab) [inline Doc/library/functions.rst]: Fix these Sphinx things

Comment (justin39) [reply Doc/library/functions.rst]: Are the links broken?

File: Objects/iterobject.c
Comment (eric-wieser) [inline Objects/iterobject.c]: I assume this comment is for reviewers, and should be removed / addressed before merging? If so, leaving this comment as a marker for someone who knows more.

File: Objects/iterobject.c
Comment (eric-wieser) [inline Objects/iterobject.c]: Shouldn't this be `AsyncStopIteration`? That seems to be how `next()` behaves on callables

File: Doc/library/functions.rst
Comment (jab) [reply Doc/library/functions.rst]: With roles like `:func:` and `:meth:`, Sphinx adds the parens at the end for you when the docs are rendered. Fixed now.  Built the html docs locally with these changes and they look good to me.

Comment (lordmauve) [reply Doc/library/functions.rst]: I think you can still include parentheses in the role text, though, right? I'm pretty sure I do that all the time, because it makes the reST more readable.

Comment (jab) [reply Doc/library/functions.rst]: I'm not sure if you can optionally include them. All the other occurrences of `:func:` in the documentation source do not include parens. At the very least, excluding them seems the preferred style.

**Comment by Fidget-Spinner** [comment]
IMO, this looks like it should be in whatsnew too under [this section](https://docs.python.org/3.10/whatsnew/3.10.html#other-language-changes). It's not everyday two new builtins get added :). Maybe something like 
```rst
* Two new builtin functions -- :func:`aiter` and :func:`anext` have been added
  to provide asynchronous counterparts to :func:`iter` and :func:`anext` 
  respectively.
```

**Review by nanjekyejoannah — COMMENTED** [review]
I agree, a whatsnew entry is required in this case. I will ping @1st1 for a look.

File: Objects/abstract.c
Comment (justin39) [inline Objects/abstract.c]: Whoops, think we're missing a null check for `tp_as_async` here

File: Objects/iterobject.c
Comment (justin39) [reply Objects/iterobject.c]: Makes sense to me - @lordmauve we don't add this error message if we raise `StopAsyncIteration` right? Formatting with `obj->it` doesn't seem to work anyways, think we would need `Py_TYPE(obj->it)->tp_name`

Comment (jab) [reply Objects/iterobject.c]: Done in 06019ea05ebcd477f1545f2fccc8b089204ade42. I think this can be resolved.

**Review by ghost — APPROVED** [review]
LGTM

**Comment by jab** [comment]
Thanks for pinging @1st1 for review, @nanjekyejoannah, and hope this finds you both well. Is there still interest in this on your end? A bot just added the “stale” label, but there is still interest on our end in seeing this through to completion. Thanks!

**Comment by jab** [comment]
Since this PR has again been automatically marked "stale" due to inactivity, leaving a comment just to say there is still interest on our end in seeing this through to completion.

Hi @1st1, could you please acknowledge whether this is on your radar? @nanjekyejoannah requested a review from you in December. I think you're in the role of primary reviewer here, and other reviewers may be waiting for your review, but it's not clear whether you saw this. Do you have interest in reviewing this at some point? If not, would you be able to suggest someone else? Thank you!

**Comment by booski360** [comment]
Wil do I’ll get right on it

**Comment by jab** [comment]
Shoot, noticed this PR now has conflicts. Will resolve and push a new revision ASAP.

**Comment by justin39** [comment]
Resolved the conflicts - hopefully we can get this reviewed before a new conflict pops up!

File: Objects/iterobject.c
Comment (jab) [reply Objects/iterobject.c]: Removed this comment in the latest revision so this PR is in a more mergeable state, but leaving this conversation unresolved until we get confirmation. The removed comment was:
> Can we raise this at this point, or do we need to return an awaitable that raises it?

...and referred to (what's currently on line 368), `PyErr_SetNone(PyExc_StopAsyncIteration);`.

**Comment by jab** [comment]
Merged in latest master and fixed some minor nits. All checks have past against the latest revision.

UPDATE (ICYMI): There is now ongoing discussion in [this recent python-dev thread](https://mail.python.org/archives/list/python-dev@python.org/thread/5XUVPB5H4PFUGTC5F7KAN4STKAEOFBQM/) about whether to merge this PR (favored by @1st1, and perhaps others who've ❤️'d and participated in this PR), or whether to instead resurrect #8895, my PR from 2018 that added `operator.aiter` and `operator.anext` (favored by @ramalho, @pbryan, @gvanrossum, @brettcannon, @terryjreedy). Looking forward to seeing what the core developers decide, and then hopefully being able to merge one of these two PRs before the upcoming 3.10 feature freeze on May 3. Thank you!

File: Doc/glossary.rst
Comment (gvanrossum) [inline Doc/glossary.rst]: I don't think we need yet another vague term in the glossary. Honestly all the "asynchronous _something_" entries in the glossary seem to be misguided, but we can discuss that another time (I literally never look anything up in the glossary, so maybe I am just not the right audience).

File: Doc/library/functions.rst
Comment (gvanrossum) [inline Doc/library/functions.rst]: Frankly, this doesn't tell me much. The description doesn't even state that `aiter(x)` is equivalent to `x.__aiter__()`, which to me is the key point. Certainly it shouldn't start by stating the type of what it returns; it should describe how the return value relates to the input. (Compare the entry for abs(x) above, "Return the absolute value of a number." This clearly references the input and what the function does to that value.)

Also, state explicitly that `aiter(aiter(x))` is the same as `aiter(x)` (IOW that `aiter(x)` itself has an `__aiter__()` method that returns `self`.)

File: Doc/library/functions.rst
Comment (gvanrossum) [inline Doc/library/functions.rst]: I don't think we need to explain this.

File: Include/abstract.h
Comment (gvanrossum) [inline Include/abstract.h:330]: Is it necessary to add these to the public C API? Just because `PyObject_GetIter()` is public I'm not sure that the Aiter variant needs to be. @vstinner tends to push back on adding new things to the C API. @1st1 what do you think?

File: Include/iterobject.h
Comment (gvanrossum) [inline Include/iterobject.h]: Maybe these types and the function below should remain CPython implementation details? Just because they're returned by builtins doesn't mean all the implementation types need to be in the C-level API.

File: Objects/abstract.c
Comment (gvanrossum) [inline Objects/abstract.c]: You don't need those backslashes, since this is not a macro.

**Review by gvanrossum — COMMENTED** [review]
The basic code is solid. Some questions and suggestions about how much needs to be public and documented.

File: Doc/glossary.rst
Comment (jab) [reply Doc/glossary.rst]: Oops, meant to delete this when we removed the 2-arg variant of `aiter()`. Removed now. Thanks!

File: Doc/library/functions.rst
Comment (jab) [reply Doc/library/functions.rst]: Incorporated in the latest revision, but please let me know if it needs further refinement. Thanks for the great feedback!

File: Doc/library/functions.rst
Comment (jab) [reply Doc/library/functions.rst]: Removed.

File: Objects/abstract.c
Comment (jab) [reply Objects/abstract.c]: Removed in the latest revision.

File: Doc/library/functions.rst
Comment (gvanrossum) [inline Doc/library/functions.rst]: Much better! Here is how I would rearrange this:
```
Return an :term:`asynchronous iterator`
for an :term:`asynchronous iterable`. 
Equivalent to calling ``x.__aiter__()``.

``aiter(x)`` itself has an ``__aiter__()`` method that returns ``x``,
so ``aiter(aiter(x))`` is the same as ``aiter(x)``.
```

**Review by gvanrossum — COMMENTED** [review]
Almost there!

File: Doc/library/functions.rst
Comment (jab) [reply Doc/library/functions.rst]: Incorporated in the latest revision, thanks!

(((
Come to think of it, `iter` also has the `iter(iter(x)) == iter(x)` property, which is not currently mentioned in [the `iter` docs](https://docs.python.org/3/library/functions.html#iter). The `iter` docs have a lot more work to do though, to cover the 1- and 2-arg variants (and they're already doing this very well).

And now that I'm looking at those again, I notice the only code example there is for the 2-arg variant:
```python
from functools import partial
with open('mydata.db', 'rb') as f:
    for block in iter(partial(f.read, 64), b''):
        process_block(block)
```
If this pattern is now obsoleted by assignment expressions[1], is it worth updating the `iter` docs to (1) mention the `iter(iter(x)) == iter(x)` property, and (2) remove the obsoleted example code? If so, happy to submit a separate PR for that.

[1] as in the following:

```python
with open('mydata.db', 'rb') as f:
    while block := f.read(64):
        process_block(block)
```
)))

File: Doc/library/functions.rst
Comment (lordmauve) [reply Doc/library/functions.rst]: One reason I quite like this snippet is that it helps to avoid surprise at why `aiter()` doesn't have a 2-argument form to match `iter()`. But it does feel out of place to talk about things this function _doesn't do_.

File: Doc/library/functions.rst
Comment (gvanrossum) [reply Doc/library/functions.rst]: I don't want to personally get into the weeds about the iter() docs, sorry. Something for the docs WG perhaps.

Comment (brettcannon) [reply Doc/library/functions.rst]: I don't think it's worth updating the docs to promise `iter(iter(x)) == iter(x)` as it isn't a specific benefit to the user beyond logically doing the right thing.

File: Include/abstract.h
Comment (1st1) [reply Include/abstract.h:330]: I don't mind adding this function -- while somewhat trivial, it's something that projects like Cython (and potentially our own modules like `_asynciomodule.c`) have to reimplement.

File: Include/iterobject.h
Comment (1st1) [reply Include/iterobject.h]: +1. I'd make them internal.

Comment (jab) [reply Include/iterobject.h]: Makes sense, thanks. Done in the latest revision. Another look?

File: Doc/library/functions.rst
Comment (jab) [reply Doc/library/functions.rst]: I can see that.

@gvanrossum, do you think @brettcannon's rationale applies equally to the `aiter` docs too? I did add a `test_aiter_idempotent()` for this in 6ee8824 to go along with the promise that the `aiter` docs are now making, but can remove that test along with that part of the docs if that's better.

Comment (gvanrossum) [reply Doc/library/functions.rst]: I think there's a different bar for updating the `iter()` docs than for the initial version of the `aiter()` docs. And they don't have to match precisely.

FWIW the main reason the idempotency property is important is because of the implicit `[a]iter()` call in a for-loop, since in
```
for i in x: ...
```
the for-loop calls `iter(x)`, so that in
```
for i in iter(x): ...
```
the for-loop ends up calling `iter(iter(x))`. It's the same for `async for` and `aiter()`.

File: Doc/library/functions.rst
Comment (gvanrossum) [inline Doc/library/functions.rst]:
```suggestion
   This calls the :meth:`~object.__anext__` method of *async_iterator*,
   returning an :term:`awaitable`. Awaiting this returns the next value of the
```

**Review by gvanrossum — APPROVED** [review]
Thanks for your patience. This all looks good to me now! I'll merge after applying my own suggestion.

**Comment by gvanrossum** [comment]
Congrats!

**Comment by jab** [comment]
Thanks so much, @gvanrossum! We’re honored to have been able to contribute!

**Comment by gvanrossum** [comment]
I believe test_asyncio is unstable or broken on this platform.

**Comment by vstinner** [comment]
I proposed to PR #25266 to rename PyAnextAwaitable_Type to _PyAnextAwaitable_Type, and to initialize the type at Python startup: can someone please have a look?

**Pull request merged**
